### ISSUE TITLE
FR: Return a sensible DS (Delegation Signer) record for MagicDNS

### ISSUE THREAD

**Issue by normanr** [tailscale/tailscale#4726]
### What is the issue?

I tried to query for `magicdns.localhost-tailscale-daemon` via bind9 (forwarding to magicdns), and the lookup fails because bind9 also queries for a `DS` record and `tailscaled` ignores the query (tries to forward it upsteam, but `upstream nameservers not set` so it fails).

I think this could be fixed by adding cases to the switch statement for `dnsSymbolicFQDN`: https://github.com/tailscale/tailscale/blob/dce2409b15837f30885405b8b1d27e7b7fd6bf7a/net/dns/resolver/tsdns.go#L630-L635
which would return `RCodeNotImplemented` or `RCodeSuccess` like https://github.com/tailscale/tailscale/blob/dce2409b15837f30885405b8b1d27e7b7fd6bf7a/net/dns/resolver/tsdns.go#L696-L712
This would prevent the code falling through to checking if the `!found` name is in `localDomains` and then returning `RCodeRefused` (which triggers the forwarding upstream).

### Steps to reproduce

Configure bind9 with:
```
zone "localhost-tailscale-daemon" { type forward; forward only; forwarders { 100.100.100.100; fd7a:115c:a1e0::53; }; };
```
and try and lookup `magicdns.localhost-tailscale-daemon` via bind9.

### OS

Linux

### OS version

Debian testing (bookworm)

### Tailscale version

1.24.2

**Comment by normanr** [comment]
heh, in the same way `dig +short ANY magicdns.localhost-tailscale-daemon @100.100.100.100` fails, because there is case for `TypeAll`

**Comment by andrew-d** [comment]
I spent some time looking into this today, and unfortunately just responding to the `DS` record type doesn't appear to work (testing with 9.18.3). This is because the lack of a `DS` record itself isn't enough to disable DNSSEC; a DNS server should explicitly send a `NSEC` or `NSEC3` record to "prove" that DNSSEC isn't turned on for a subdomain. In the case of BIND, it doesn't have any way to get one of these for the `localhost-tailscale-daemon` zone, and thus assumes that something is tampering with responses and stripping DNSSEC, and fails.

As a workaround, you can explicitly tell BIND that this domain should not have DNSSEC in the configuration file; something like this worked for me:
```
options {
    dnssec-validation yes;                                                                                                                                                                                                                             
    dnssec-must-be-secure "localhost-tailscale-daemon" false;
};

zone "localhost-tailscale-daemon" {
    type forward;
    forward only;
    forwarders {
        100.100.100.100;
        fd7a:115c:a1e0::53;
    };  
};
```

When I run it (with `listen-on port 5354 { 127.0.0.1; };` for testing), I can then do this:
```
$ dig A magicdns.localhost-tailscale-daemon @127.0.0.1 -p 5354

; <<>> DiG 9.18.3 <<>> A magicdns.localhost-tailscale-daemon @127.0.0.1 -p 5354
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 64923
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; COOKIE: 941927a768ed932001000000635c361d44d3705e63be7f7e (good)
;; QUESTION SECTION:
;magicdns.localhost-tailscale-daemon. IN	A

;; ANSWER SECTION:
magicdns.localhost-tailscale-daemon. 600 IN A	100.100.100.100

;; Query time: 0 msec
;; SERVER: 127.0.0.1#5354(127.0.0.1) (UDP)
;; WHEN: Fri Oct 28 16:05:49 EDT 2022
;; MSG SIZE  rcvd: 108
```

I think it's unlikely that we're going to be implementing support for returning DNSSEC records from `tailscaled` in the near future, so this workaround is likely to be your best bet for now.

**Comment by normanr** [comment]
I don't think tailscale can/should return a `DS` record. It should just return NotImplemented (as if it didn't understand the query), or  Success with no results (as if the record is missing).

Tailscale can't provide a `NSEC` or `NSEC3`, but it also shouldn't just ignore requests that it should be authoritative for. It should return _something_ (either an error, or empty response).

**Comment by normanr** [comment]
ahh, actually since https://github.com/tailscale/tailscale/pull/4863, tailscale at least returns a failure (which is good).  Unfortunately returning a `SERVFAIL` for `magicdns.localhost-tailscale-daemon`, makes bind refuse to return any _other_ records for the same name.

As I said in the opening bug: tailscale should not try and forward requests for `dnsSymbolicFQDN` to upstream resolvers. It should return `NotImplemented` or `Success` (like other names that it is authoritative for). i.e. `dnsSymbolicFQDN` should be considered `found`, and _not_ return `Refused`.

**Comment by andrew-d** [comment]
At least in my testing, that didn't work in either case. With this diff:
```patch
diff --git net/dns/resolver/tsdns.go net/dns/resolver/tsdns.go
index e6a0b366..f95a53a6 100644
--- net/dns/resolver/tsdns.go
+++ net/dns/resolver/tsdns.go
@@ -548,6 +548,17 @@ func (r *Resolver) resolveLocal(domain dnsname.FQDN, typ dns.Type) (netip.Addr,
                        return tsaddr.TailscaleServiceIP(), dns.RCodeSuccess
                case dns.TypeAAAA:
                        return tsaddr.TailscaleServiceIPv6(), dns.RCodeSuccess
+
+               // We don't support DNSSEC
+               case 43 /* dns.TypeDS */ :
+                       return netip.Addr{}, dns.RCodeSuccess
+
+               // All other record types for our symbolic domain are not
+               // supported; mimic the switch statement below and return no
+               // records.
+               default:
+                       metricDNSResolveNoRecordType.Add(1)
+                       return netip.Addr{}, dns.RCodeSuccess
                }
        }
        // Special-case: 'via-<siteid>.<ipv4>' queries.
```

BIND still errors; here's the selected lines:
<details>
  <summary>Click to expand the log</summary>

```
28-Oct-2022 18:00:54.863 clientmgr @0x7fc2a00c7100 attach: 2
28-Oct-2022 18:00:54.863 client @0x7fc29821e168 (no-peer): allocate new client
28-Oct-2022 18:00:54.864 client @0x7fc29821e168 127.0.0.1#44854: UDP request
28-Oct-2022 18:00:54.864 client @0x7fc29821e168 127.0.0.1#44854: using view '_default'
28-Oct-2022 18:00:54.864 client @0x7fc29821e168 127.0.0.1#44854: request is not signed
28-Oct-2022 18:00:54.864 client @0x7fc29821e168 127.0.0.1#44854: recursion available
28-Oct-2022 18:00:54.864 client @0x7fc29821e168 127.0.0.1#44854 (magicdns.localhost-tailscale-daemon): query (cache) 'magicdns.localhost-tailscale-daemon/A/IN' approved
28-Oct-2022 18:00:54.864 fetch: magicdns.localhost-tailscale-daemon/A
28-Oct-2022 18:00:54.864 log_ns_ttl: fctx 0x7fc298243000: fctx_create: magicdns.localhost-tailscale-daemon (in '.'?): 1 518386
28-Oct-2022 18:00:54.864 QNAME minimization -  minimized, qmintype 1 qminname _.localhost-tailscale-daemon
28-Oct-2022 18:00:54.865 received packet from 100.100.100.100#53
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id:  18623
;; flags: qr aa rd ra cd; QUESTION: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; QUESTION SECTION:
;magicdns.localhost-tailscale-daemon. IN        A

;; ANSWER SECTION:
;magicdns.localhost-tailscale-daemon. 600 IN A  100.100.100.100


28-Oct-2022 18:00:54.865 received packet (no opt) from 100.100.100.100#53
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id:  18623
;; flags: qr aa rd ra cd; QUESTION: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; QUESTION SECTION:
;magicdns.localhost-tailscale-daemon. IN        A

;; ANSWER SECTION:
magicdns.localhost-tailscale-daemon. 600 IN A   100.100.100.100


28-Oct-2022 18:00:54.865 log_ns_ttl: fctx 0x7fc298243000: rctx_answer: magicdns.localhost-tailscale-daemon (in 'localhost-tailscale-daemon'?): 1 518386
28-Oct-2022 18:00:54.865 delete_node(): 0x7fc277400140 magicdns.localhost-tailscale-daemon (bucket 14)
28-Oct-2022 18:00:54.865 validating magicdns.localhost-tailscale-daemon/A: starting
28-Oct-2022 18:00:54.865 validating magicdns.localhost-tailscale-daemon/A: attempting insecurity proof
28-Oct-2022 18:00:54.865 validating magicdns.localhost-tailscale-daemon/A: checking existence of DS at 'localhost-tailscale-daemon'
28-Oct-2022 18:00:54.865 validating magicdns.localhost-tailscale-daemon/A: proveunsecure: creating fetch for localhost-tailscale-daemon DS
28-Oct-2022 18:00:54.865 fetch: localhost-tailscale-daemon/DS
[...]
28-Oct-2022 18:00:54.891 received packet from 192.5.5.241#53
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id:  28411
;; flags: qr aa; QUESTION: 1, ANSWER: 0, AUTHORITY: 6, ADDITIONAL: 1
;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags: do; udp: 1232
; COOKIE: 57d4cf07fb9151e601000000635c51166e6f9ba4bf998e94
;; QUESTION SECTION:
;localhost-tailscale-daemon.    IN      DS

;; AUTHORITY SECTION:
;.                      86400   IN      SOA     a.root-servers.net. nstld.verisign-grs.com. (
;                                               2022102801 ; serial
;                                               1800       ; refresh (30 minutes)
;                                               900        ; retry (15 minutes)
;                                               604800     ; expire (1 week)
;                                               86400      ; minimum (1 day)
;                                               )
;.                      86400   IN      RRSIG   SOA 8 0 86400 (
;                                               20221110170000 20221028160000 18733 .
;                                               trhMh6R/gdm8Yzpl+gUj/y/zaQkg
;                                               o8AcTUoX9XH6xvNDbM72ei2bDZht
;                                               hziGQLeZxs/cFk14huRINu6xzRbF
;                                               SDqba2ju31niy1/IigqQRa0xWwbR
;                                               9PfURx1BkLLUndlYytS+/5fpQJjo
;                                               T6NcPfYX1UIZEFTaofmmPHtmdpzE
;                                               8oiJqYQ5Fu05UfaBrWZTjVPsgO3i
;                                               2zw0d236DhKedX1/eUm5iNEtB+Fh
;                                               7vTW+8C0FZQi8tA02udRQjvO5m2R
;                                               sVnunlIhznC9p4gUEddt4Q4T3KAu
;                                               w7Ziex6Iw6f7OysbsFqGkNhDQsHi
;                                               tknVbj+nQF/rXGAZJYtOVRvJpASt
;                                               E4cc6Q== )
;.                      86400   IN      NSEC    aaa. NS SOA RRSIG NSEC DNSKEY
;.                      86400   IN      RRSIG   NSEC 8 0 86400 (
;                                               20221110170000 20221028160000 18733 .
;                                               NDvswSjzKLhdkDKWjZiHemHMa/6y
;                                               klq//XxL47u4Cp7vdgZjVCvjgZH3
;                                               63YG0g/LKAB2j9m+q8f08lw6oJnz
;                                               UrLP0Ug/26W/FcLm1ojabXUjHYph
;                                               q4w2C+EVMoNrIiNNKnvWGpesuYd4
;                                               iTtzidbk5FCBu9zxb4Qj+jrQe+s8
;                                               hJdg4kCpE05CHGF9d63OlMi6TABy
;                                               bzxfO2e2I+kr5C6XU5Pwk70fi7Ek
;                                               +g0gotsD/VRTl9mQm6UTjb5ZcuqH
;                                               dvXhOhXkgKHIbONHaojSc3Sd2bq/
;                                               kX8SXO8Vg6p4OmnY+S83/HVmJwCt
;                                               izbG2hGa8x1IzTZ/cWIyJeBqgy64
;                                               75n9Bg== )
;loans.                 86400   IN      NSEC    locker. NS DS RRSIG NSEC
;loans.                 86400   IN      RRSIG   NSEC 8 1 86400 (
;                                               20221110170000 20221028160000 18733 .
;                                               3//XGhMxGfLHKtKs3O53nQZ+zqOv
;                                               up4TOWOx45CFSrjNGT5DStT/4REJ
;                                               zfNJzb53q+jmpHKofnnO6cxEFnkn
;                                               RsnVEau0Jd6N6IqG7ZESV7LoOHL9
;                                               ou1SDetDhn61S+MfO328bAYbqQpb
;                                               B4N4+I9TEowk5kcFImr0ych2MFvn
;                                               pP8IJiURFASvjoiJe2QPXfL7vz+t
;                                               ip6AgyCM8uWcXQqe8ePQFSJL5fba
;                                               hpcCQzMSGHWmOlHvhLyVKCks00/D
;                                               /EHUgQGnc8t2MdIYfbLYTJfgHmkD
;                                               8mwAigtGNL/OUArcHkI//WK5wqBc
;                                               2lYyie12fqahYle/u4kAJrSeVDOx
;                                               NbbvAg== )
[...]
28-Oct-2022 18:00:54.891 log_ns_ttl: fctx 0x7fc2a0138c00: rctx_answer_none: localhost-tailscale-daemon (in '.'?): 1 518386
28-Oct-2022 18:00:54.891 validating localhost-tailscale-daemon/DS: starting
28-Oct-2022 18:00:54.891 validating localhost-tailscale-daemon/DS: attempting negative response validation from message
28-Oct-2022 18:00:54.891 validating localhost-tailscale-daemon/DS: validate_neg_rrset: creating validator for . SOA
28-Oct-2022 18:00:54.891   validating ./SOA: starting
28-Oct-2022 18:00:54.891   validating ./SOA: attempting positive response validation
28-Oct-2022 18:00:54.891   validating ./SOA: keyset with trust secure
28-Oct-2022 18:00:54.892   validating ./SOA: verify rdataset (keyid=18733): success
28-Oct-2022 18:00:54.892   validating ./SOA: marking as secure, noqname proof not needed
28-Oct-2022 18:00:54.892   validator @0x7fc277e2c400: dns_validator_destroy
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: in validator_callback_nsec
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: resuming validate_nx
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: validate_neg_rrset: creating validator for . NSEC
28-Oct-2022 18:00:54.892   validating ./NSEC: starting
28-Oct-2022 18:00:54.892   validating ./NSEC: attempting positive response validation
28-Oct-2022 18:00:54.892   validating ./NSEC: keyset with trust secure
28-Oct-2022 18:00:54.892   validating ./NSEC: verify rdataset (keyid=18733): success
28-Oct-2022 18:00:54.892   validating ./NSEC: marking as secure, noqname proof not needed
28-Oct-2022 18:00:54.892   validator @0x7fc277e2c400: dns_validator_destroy
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: in validator_callback_nsec
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: looking for relevant NSEC
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: ignoring nsec because name is past end of range
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: resuming validate_nx
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: validate_neg_rrset: creating validator for loans NSEC
28-Oct-2022 18:00:54.892   validating loans/NSEC: starting
28-Oct-2022 18:00:54.892   validating loans/NSEC: attempting positive response validation
28-Oct-2022 18:00:54.892   validating loans/NSEC: keyset with trust secure
28-Oct-2022 18:00:54.892   validating loans/NSEC: verify rdataset (keyid=18733): success
28-Oct-2022 18:00:54.892   validating loans/NSEC: marking as secure, noqname proof not needed
28-Oct-2022 18:00:54.892   validator @0x7fc277e2c400: dns_validator_destroy
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: in validator_callback_nsec
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: looking for relevant NSEC
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: nsec range ok
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: resuming validate_nx
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: in checkwildcard: *
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: looking for relevant NSEC
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: nsec range ok
28-Oct-2022 18:00:54.892 validating localhost-tailscale-daemon/DS: nonexistence proof(s) found
[...]
28-Oct-2022 18:00:54.892 log_ns_ttl: fctx 0x7fc298ec9000: fctx_create: magicdns.localhost-tailscale-daemon (in 'localhost-tailscale-daemon'?): 0 0
28-Oct-2022 18:00:54.893 received packet from 100.100.100.100#53
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id:   5475
;; flags: qr aa rd ra; QUESTION: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0
;; QUESTION SECTION:
;magicdns.localhost-tailscale-daemon. IN        DS


28-Oct-2022 18:00:54.893 log_ns_ttl: fctx 0x7fc298ec9000: rctx_answer_none: magicdns.localhost-tailscale-daemon (in 'localhost-tailscale-daemon'?): 0 0
28-Oct-2022 18:00:54.893 validating magicdns.localhost-tailscale-daemon/DS: starting
28-Oct-2022 18:00:54.893 validating magicdns.localhost-tailscale-daemon/DS: attempting negative response validation from message
28-Oct-2022 18:00:54.893 validating magicdns.localhost-tailscale-daemon/DS: nonexistence proof(s) not found
28-Oct-2022 18:00:54.893 validating magicdns.localhost-tailscale-daemon/DS: checking existence of DS at 'localhost-tailscale-daemon'
28-Oct-2022 18:00:54.893 validating magicdns.localhost-tailscale-daemon/DS: checking existence of DS at 'magicdns.localhost-tailscale-daemon'
28-Oct-2022 18:00:54.893 validating magicdns.localhost-tailscale-daemon/DS: continuing validation would lead to deadlock: aborting validation
28-Oct-2022 18:00:54.893 validating magicdns.localhost-tailscale-daemon/DS: deadlock found (create_fetch)
28-Oct-2022 18:00:54.893 validator @0x7fc2a00a7c00: dns_validator_destroy
28-Oct-2022 18:00:54.893 no valid RRSIG resolving 'magicdns.localhost-tailscale-daemon/DS/IN': 100.100.100.100#53
28-Oct-2022 18:00:54.893 validating magicdns.localhost-tailscale-daemon/A: in fetch_callback_ds
28-Oct-2022 18:00:54.893 validating magicdns.localhost-tailscale-daemon/A: fetch_callback_ds: got SERVFAIL
28-Oct-2022 18:00:54.893 validator @0x7fc2a00a7200: dns_validator_destroy
28-Oct-2022 18:00:54.893 broken trust chain resolving 'magicdns.localhost-tailscale-daemon/A/IN': 100.100.100.100#53
28-Oct-2022 18:00:54.893 client @0x7fc29821e168 127.0.0.1#44854 (magicdns.localhost-tailscale-daemon): query failed (broken trust chain) for magicdns.localhost-tailscale-daemon/IN/A at query.c:7715
28-Oct-2022 18:00:54.893 fetch completed at resolver.c:5601 for magicdns.localhost-tailscale-daemon/A in 0.028999: broken trust chain/broken trust chain [domain:localhost-tailscale-daemon,referral:0,restart:1,qrysent:1,timeout:0,lame:0,quota:0,neterr:0,badresp:0,adberr:0,findfail:0,valfail:1]
28-Oct-2022 18:00:54.893 client @0x7fc29821e168 127.0.0.1#44854 (magicdns.localhost-tailscale-daemon): reset client
```
</details>

The same error occurs if we return `dns.RCodeNameError` (`NXDOMAIN`), or `dns.RCodeNotImplemented`.

Based on my read of the log--with the caveat that I'm not a DNSSEC expert--what's happening is that BIND is reaching out to the root servers, getting a NSEC response for non-existence of the `localhost-tailscale-daemon/DS` record (`nonexistence proof(s) found`). Then, it attempts to prove that there's no `DS` record for `magicdns.localhost-tailscale-daemon` (`checking existence of DS at 'magicdns.localhost-tailscale-daemon'`), and cannot do so because the Tailscale DNS server does not return a `NSEC` or `NSEC3` record (`nonexistence proof(s) not found`).

**Comment by andrew-d** [comment]
> From https://groups.google.com/g/comp.protocols.dns.bind/c/D1f00tOQE0o (and other places) it looks like this is not the _intended_ use for the `dnssec-must-be-secure` keyword. `validate-except` seem to be a more appropriate option ([available since bind 9.13](https://www.isc.org/blogs/bind-9-new-versions/)).

Yep, that looks more correct, and I can confirm that it works when testing locally. Nice find!
